server/src/main/java/org/elasticsearch/index/mapper/DocValueFetcher.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.mapper;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.search.DocValueFormat;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Value fetcher that loads from doc values.
+ */
+public final class DocValueFetcher implements ValueFetcher {
+    private final DocValueFormat format;
+    private final IndexFieldData<?> ifd;
+    private Leaf leaf;
+
+    public DocValueFetcher(DocValueFormat format, IndexFieldData<?> ifd) {
+        this.format = format;
+        this.ifd = ifd;
+    }
+
+    public void setNextReader(LeafReaderContext context) {
+        leaf = ifd.load(context).getLeafValueFetcher(format);
+    }
+
+    @Override
+    public List<Object> fetchValues(int docId) throws IOException {
+        if (false == leaf.advanceExact(docId)) {
+            return List.of();
+        }
+        List<Object> result = new ArrayList<Object>(leaf.docValueCount());
+        for (int i = 0, count = leaf.docValueCount(); i < count; ++i) {
+            result.add(leaf.nextValue());
+        }
+        return result;
+    }
+
+    public interface Leaf {
+        /**
+         * Advance the doc values reader to the provided doc.
+         * @return false if there are no values for this document, true otherwise
+         */
+        boolean advanceExact(int docId) throws IOException;
+
+        /**
+         * A count of the number of values at this document.
+         */
+        int docValueCount() throws IOException;
+
+        /**
+         * Load and format the next value.
+         */
+        Object nextValue() throws IOException;
+    }
```

Comment:
These methods look suspiciously similar (the same?) to those exposed by all of our Sorted*DocValues implementations. I have been thinking about why we need this new interface, given that conceptually, what we are doing when fetching values from doc_values is exactly the same as usual (reading!) plus some reformatting on what nextValue returns.

I think that we have a chance to make the integration with the existing doc values classes easier to follow. I am under the impression that we need this new interface mostly because all of our Sorted*DocValues classes don't share a base class, is that the case? Would it be possible / make sense to have LeafFieldData instead return an enhanced version of Sorted*DocValues that does the formatting, rather than introducing a new interface for that that has the same methods as the existing classes?

This may become a bigger change, I haven't thought all of this through, but maybe it makes sense to discuss as we may have a chance to reduce the boilerplate we need to add here.

Reply:
> _Would it be possible / make sense to have LeafFieldData instead return an enhanced version of Sorted_DocValues that does the formatting, rather than introducing a new interface for that that has the same methods as the existing classes?

That's a great idea! I hadn't thought of it. And now that I look at it maybe it won't work. I think the return of `nextValue` has to line up - `SortedBinaryDocValues` returns `BytesRef`. `SortedNumericDoubleValues` returns `double`. `SortedNumericDocValues` returns `long`. And It has a bunch of extra methods it inherits from Lucene that don't make sense here. So to reuse those in this context we'd want to make one that returns `Object` and wrap them. Which is what this ended up being. Sort of on accident.

I do wonder if it'd be useful to make a common interface that declares `advanceExact` and `docValueCount`. I don't know that it'd save a ton of work though.

Reply:
I spent some time looking at alternatives but given that getValues return a different type, we don't want auto-boxing, and DocValueFormat also has specific format methods per type, I don't think this puzzle can be solved better than you already did. Looking at it from that perspective, what you achieved is pretty nice!